uvicorn/_async_agnostic/server.py

```diff
@@ -0,0 +1,268 @@
+import contextlib
+import logging
+import os
+import signal
+import socket
+import threading
+import time
+from typing import Iterator, List, Optional
+
+import click
+
+from ..config import Config
+from .backends.auto import select_async_backend
+from .backends.base import AsyncListener, Event, TaskStatus
+from .http11.handler import handle_http11
+from .lifespan import Lifespan
+from .state import ServerState
+from .utils import to_internet_date
+
+HANDLED_SIGNALS = (
+    signal.SIGINT,  # Unix signal 2. Sent by Ctrl+C.
+    signal.SIGTERM,  # Unix signal 15. Sent by `kill <pid>`.
+)
+
+
+class Server:
+    def __init__(self, config: Config) -> None:
+        self._config = config
+        self._state = ServerState()
+        self._backend = select_async_backend(config.async_library)
+        self._logger = logging.getLogger("uvicorn.error")
+        self._last_notified = time.time()
+        self._force_exit = False
+        self._sync_started_event = threading.Event()
+
+    @property
+    def _shutdown_event(self) -> Event:
+        if not hasattr(self, "_shutdown_event_obj"):
+            # Can't be created on init due to a limitation of multiprocessing.
+            self._shutdown_event_obj = self._backend.create_event()
+        return self._shutdown_event_obj
+
+    @property
+    def _closed_event(self) -> Event:
+        if not hasattr(self, "_closed_event_obj"):
+            # Can't be created on init due to a limitation of multiprocessing.
+            self._closed_event_obj = self._backend.create_event()
+        return self._closed_event_obj
+
+    def run(self, sockets: List[socket.SocketType] = None) -> None:
+        if self._config.async_library == "asyncio":
+            self._config.setup_event_loop()
+        self._backend.run(self._main, sockets)
+
+    @contextlib.contextmanager
+    def run_in_thread(self, sockets: List[socket.SocketType] = None) -> Iterator[None]:
```

Comment:
This general-purpose helper function is something people end up having to re-implement anytime they want to use Uvicorn in a tests setup. (We do it ourselves, we do it in HTTPX, in HTTPCore, I do it in my projects, etc.) We could consider adding this to the existing `Server` implementation, if we'd like.